File: solutions/1116-maximum-level-sum-of-a-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLevelSum(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int maxsum = INT_MIN;
        int level  = 1;
        int maxlevel = 1;
        if(!root)return 0;

        while(!q.empty()){
            int sum = 0;
            int n= q.size();
            for(int i=0;i<n;i++){
                TreeNode* curr = q.front();q.pop();
                sum+= curr->val;
                if(curr->left){
                    q.push(curr->left);
                }
                if(curr->right){
                    q.push(curr->right);
                }
            }
            if(maxsum<sum){
                maxsum = sum;
                maxlevel = level;
            }
            level++;
        }
        
        return maxlevel;
    }
};

File: solutions/1116-maximum-level-sum-of-a-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        # level order traversal and keep track of max sum of level and the level
        maxsumlevel = 1
        maxsum = root.val
        q = [root]
        level = 1
        while len(q)!=0:
            newq = []
            currsum = 0
            for i in range(0,len(q)):
                curr = q[i]
                if curr.left:
                    newq.append(curr.left)
                if curr.right:
                    newq.append(curr.right)
                currsum+= curr.val
            q = newq
            if maxsum<currsum:
                maxsum = currsum
                maxsumlevel = level
            level+=1
        return maxsumlevel





        
